### PR TITLE
[BugFix] Dangling ptr in schema change

### LINKED ISSUE #6123: stack-use-after-scope in asan mode.
### Steps to reproduce the behavior (Required)
1. CREATE TABLE '...'
2. INSERT INTO '....'
3. SELECT  '....'

### Real behavior (Required)
```
22/05/02 13:13:22 INFO hikari.HikariDataSource: HikariPool-1 - Shutdown completed.
=================================================================
==3915==ERROR: AddressSanitizer: stack-use-after-scope on address 0x7f5ac0d5b090 at pc 0x000006d6039c bp 0x7f5ac0d56d90 sp 0x7
f5ac0d56d88
READ of size 4 at 0x7f5ac0d5b090 thread T515
    #0 0x6d6039b in starrocks::HyperLogLog::HyperLogLog(starrocks::HyperLogLog const&) /root/starrocks/be/src/storage/hll.h:93
    #1 0x6d7f32e in void __gnu_cxx::new_allocator<starrocks::HyperLogLog>::construct<starrocks::HyperLogLog, starrocks::HyperL
ogLog const&>(starrocks::HyperLogLog*, starrocks::HyperLogLog const&) /usr/include/c++/10.3.0/ext/new_allocator.h:150
    #2 0x6d76101 in void std::allocator_traits<std::allocator<starrocks::HyperLogLog> >::construct<starrocks::HyperLogLog, sta
rrocks::HyperLogLog const&>(std::allocator<starrocks::HyperLogLog>&, starrocks::HyperLogLog*, starrocks::HyperLogLog const&) /
usr/include/c++/10.3.0/bits/alloc_traits.h:512
    #3 0x6d72061 in starrocks::HyperLogLog& std::vector<starrocks::HyperLogLog, std::allocator<starrocks::HyperLogLog> >::empl
ace_back<starrocks::HyperLogLog const&>(starrocks::HyperLogLog const&) /usr/include/c++/10.3.0/bits/vector.tcc:115
    #4 0x6d640a2 in starrocks::vectorized::ObjectColumn<starrocks::HyperLogLog>::append(starrocks::HyperLogLog const*) /root/s
tarrocks/be/src/column/object_column.cpp:47
    #5 0x4f21b3b in starrocks::vectorized::ObjectColumn<starrocks::HyperLogLog>::append_datum(starrocks::vectorized::Datum con
st&) /root/starrocks/be/src/column/object_column.h:83
    #6 0x6d48a5d in starrocks::vectorized::NullableColumn::append_datum(starrocks::vectorized::Datum const&) /root/starrocks/b
e/src/column/nullable_column.cpp:45
    #7 0x4fd07e3 in starrocks::vectorized::ChunkChanger::change_chunkV2(std::shared_ptr<starrocks::vectorized::Chunk>&, std::s
hared_ptr<starrocks::vectorized::Chunk>&, starrocks::vectorized::Schema const&, starrocks::vectorized::Schema const&, starrock
s::MemPool*) /root/starrocks/be/src/storage/schema_change.cpp:593
    #8 0x4fd9994 in starrocks::vectorized::SchemaChangeDirectly::processV2(starrocks::vectorized::TabletReader*, starrocks::Ro
wsetWriter*, std::shared_ptr<starrocks::Tablet>, std::shared_ptr<starrocks::Tablet>, std::shared_ptr<starrocks::Rowset>) /root
/starrocks/be/src/storage/schema_change.cpp:948
    #9 0x4fea481 in starrocks::vectorized::SchemaChangeHandler::_convert_historical_rowsets(starrocks::vectorized::SchemaChang
eHandler::SchemaChangeParams&) /root/starrocks/be/src/storage/schema_change.cpp:1558
    #10 0x4fe6cad in starrocks::vectorized::SchemaChangeHandler::_do_process_alter_tablet_v2_normal(starrocks::TAlterTabletReq
V2 const&, starrocks::vectorized::SchemaChangeHandler::SchemaChangeParams&, std::shared_ptr<starrocks::Tablet> const&, std::sh
ared_ptr<starrocks::Tablet> const&) /root/starrocks/be/src/storage/schema_change.cpp:1448
    #11 0x4fe40a8 in starrocks::vectorized::SchemaChangeHandler::_do_process_alter_tablet_v2(starrocks::TAlterTabletReqV2 cons
t&) /root/starrocks/be/src/storage/schema_change.cpp:1349
    #12 0x4fe1a09 in starrocks::vectorized::SchemaChangeHandler::process_alter_tablet_v2(starrocks::TAlterTabletReqV2 const&)
/root/starrocks/be/src/storage/schema_change.cpp:1237
    #13 0x9bbccf2 in starrocks::EngineAlterTabletTask::execute() /root/starrocks/be/src/storage/task/engine_alter_tablet_task.
cpp:52
    #14 0x4b02de6 in starrocks::StorageEngine::execute_task(starrocks::EngineTask*) /root/starrocks/be/src/storage/storage_eng
ine.cpp:1060
    #15 0x6cc5e3b in starrocks::TaskWorkerPool::_alter_tablet(starrocks::TaskWorkerPool*, starrocks::TAgentTaskRequest const&,
 long, starrocks::TTaskType::type, starrocks::TFinishTaskRequest*) /root/starrocks/be/src/agent/task_worker_pool.cpp:523
    #16 0x6cc551e in starrocks::TaskWorkerPool::_alter_tablet_worker_thread_callback(void*) /root/starrocks/be/src/agent/task_
worker_pool.cpp:478
    #17 0x6cf0ba1 in void* std::__invoke_impl<void*, void* (*)(void*), starrocks::TaskWorkerPool*>(std::__invoke_other, void*
(*&&)(void*), starrocks::TaskWorkerPool*&&) /usr/include/c++/10.3.0/bits/invoke.h:60
    #18 0x6cf0ac9 in std::__invoke_result<void* (*)(void*), starrocks::TaskWorkerPool*>::type std::__invoke<void* (*)(void*),
starrocks::TaskWorkerPool*>(void* (*&&)(void*), starrocks::TaskWorkerPool*&&) /usr/include/c++/10.3.0/bits/invoke.h:95
    #19 0x6cf0a3a in void* std::thread::_Invoker<std::tuple<void* (*)(void*), starrocks::TaskWorkerPool*> >::_M_invoke<0ul, 1u
l>(std::_Index_tuple<0ul, 1ul>) /usr/include/c++/10.3.0/thread:264
    #20 0x6cf09f5 in std::thread::_Invoker<std::tuple<void* (*)(void*), starrocks::TaskWorkerPool*> >::operator()() (/home/dis
k1/sr/master_asan/be/lib/starrocks_be+0x6cf09f5)
    #21 0x6cf08df in std::thread::_State_impl<std::thread::_Invoker<std::tuple<void* (*)(void*), starrocks::TaskWorkerPool*> >
 >::_M_run() (/home/disk1/sr/master_asan/be/lib/starrocks_be+0x6cf08df)
    #22 0xca19e0f in execute_native_thread_routine ../../../.././libstdc++-v3/src/c++11/[thread.cc:80](http://thread.cc/)
    #23 0x7f5bbcdd6ea4 in start_thread (/lib64/libpthread.so.0+0x7ea4)
    #24 0x7f5bbc3f18dc in __clone (/lib64/libc.so.6+0xfe8dc)

Address 0x7f5ac0d5b090 is located in stack of thread T515 at offset 14048 in frame
    #0 0x4fb9505 in starrocks::vectorized::ChunkChanger::change_chunkV2(std::shared_ptr<starrocks::vectorized::Chunk>&, std::s
hared_ptr<starrocks::vectorized::Chunk>&, starrocks::vectorized::Schema const&, starrocks::vectorized::Schema const&, starrock
s::MemPool*) /root/starrocks/be/src/storage/schema_change.cpp:457

  This frame has 195 object(s):
    [32, 33) '<unknown>'
    [48, 56) 'st' (line 483)
    [80, 88) 'st' (line 513)
    [112, 120) 'st' (line 585)
    [144, 160) '<unknown>'
    [176, 192) '<unknown>'
    [208, 224) '<unknown>'
    [240, 256) '<unknown>'
    [272, 288) '<unknown>'
    [304, 320) '<unknown>'
    [336, 352) '<unknown>'
    [368, 384) '<unknown>'
    [400, 416) '<unknown>'
    [432, 448) '<unknown>'
    [464, 480) '<unknown>'
    [496, 512) '<unknown>'
    [528, 544) '<unknown>'
    [560, 576) '<unknown>'
    [592, 608) '<unknown>'
    [624, 640) '<unknown>'
    [656, 672) '<unknown>'
    [688, 704) 'new_type_info' (line 559)
    [720, 736) '<unknown>'
    [752, 768) '<unknown>'
    [784, 800) '<unknown>'
    [816, 832) '<unknown>'
    [848, 864) '<unknown>'
    [880, 912) '<unknown>'
    [944, 976) 'tmp' (line 561)
    [1008, 1056) 'base_datum' (line 511)
    [1088, 1136) 'new_datum' (line 512)
    [1168, 1216) 'base_datum' (line 526)
    [1248, 1296) 'new_datum' (line 526)
    [1328, 1376) 'base_datum' (line 526)
    [1408, 1456) 'new_datum' (line 526)
    [1488, 1536) 'base_datum' (line 526)
    [1568, 1616) 'new_datum' (line 526)
    [1648, 1696) 'base_datum' (line 526)
    [1728, 1776) 'new_datum' (line 526)
    [1808, 1856) 'base_datum' (line 526)
    [1888, 1936) 'new_datum' (line 526)
    [1968, 2016) 'base_datum' (line 526)
    [2048, 2096) 'new_datum' (line 526)
    [2128, 2176) 'base_datum' (line 526)
    [2208, 2256) 'new_datum' (line 526)
    [2288, 2336) 'base_datum' (line 526)
    [2368, 2416) 'new_datum' (line 526)
    [2448, 2496) 'base_datum' (line 526)
    [2528, 2576) 'new_datum' (line 526)
    [2608, 2656) 'base_datum' (line 526)
    [2688, 2736) 'new_datum' (line 526)
    [2768, 2816) 'base_datum' (line 528)
    [2848, 2896) 'new_datum' (line 528)
    [2928, 2976) 'base_datum' (line 528)
    [3008, 3056) 'new_datum' (line 528)
    [3088, 3136) 'base_datum' (line 528)
    [3168, 3216) 'new_datum' (line 528)
    [3248, 3296) 'base_datum' (line 528)
    [3328, 3376) 'new_datum' (line 528)
    [3408, 3456) 'base_datum' (line 528)
    [3488, 3536) 'new_datum' (line 528)
    [3568, 3616) 'base_datum' (line 528)
    [3648, 3696) 'new_datum' (line 528)
    [3728, 3776) 'base_datum' (line 528)
    [3808, 3856) 'new_datum' (line 528)
    [3888, 3936) 'base_datum' (line 528)
    [3968, 4016) 'new_datum' (line 528)
    [4048, 4096) 'base_datum' (line 528)
    [4128, 4176) 'new_datum' (line 528)
    [4208, 4256) 'base_datum' (line 528)
    [4288, 4336) 'new_datum' (line 528)
    [4368, 4416) 'base_datum' (line 530)
    [4448, 4496) 'new_datum' (line 530)
    [4528, 4576) 'base_datum' (line 530)
    [4608, 4656) 'new_datum' (line 530)
    [4688, 4736) 'base_datum' (line 530)
    [4768, 4816) 'new_datum' (line 530)
    [4848, 4896) 'base_datum' (line 530)
    [4928, 4976) 'new_datum' (line 530)
    [5008, 5056) 'base_datum' (line 530)
    [5088, 5136) 'new_datum' (line 530)
    [5168, 5216) 'base_datum' (line 530)
    [5248, 5296) 'new_datum' (line 530)
    [5328, 5376) 'base_datum' (line 530)
    [5408, 5456) 'new_datum' (line 530)
    [5488, 5536) 'base_datum' (line 530)
    [5568, 5616) 'new_datum' (line 530)
    [5648, 5696) 'base_datum' (line 530)
    [5728, 5776) 'new_datum' (line 530)
    [5808, 5856) 'base_datum' (line 530)
    [5888, 5936) 'new_datum' (line 530)
    [5968, 6016) 'base_datum' (line 532)
    [6048, 6096) 'new_datum' (line 532)
    [6128, 6176) 'base_datum' (line 532)
    [6208, 6256) 'new_datum' (line 532)
    [6288, 6336) 'base_datum' (line 532)
    [6368, 6416) 'new_datum' (line 532)
    [6448, 6496) 'base_datum' (line 532)
    [6528, 6576) 'new_datum' (line 532)
    [6608, 6656) 'base_datum' (line 532)
    [6688, 6736) 'new_datum' (line 532)
    [6768, 6816) 'base_datum' (line 532)
    [6848, 6896) 'new_datum' (line 532)
    [6928, 6976) 'base_datum' (line 532)
    [7008, 7056) 'new_datum' (line 532)
    [7088, 7136) 'base_datum' (line 532)
    [7168, 7216) 'new_datum' (line 532)
    [7248, 7296) 'base_datum' (line 532)
    [7328, 7376) 'new_datum' (line 532)
    [7408, 7456) 'base_datum' (line 532)
    [7488, 7536) 'new_datum' (line 532)
    [7568, 7616) 'base_datum' (line 534)
    [7648, 7696) 'new_datum' (line 534)
    [7728, 7776) 'base_datum' (line 534)
    [7808, 7856) 'new_datum' (line 534)
    [7888, 7936) 'base_datum' (line 534)
    [7968, 8016) 'new_datum' (line 534)
    [8048, 8096) 'base_datum' (line 534)
    [8128, 8176) 'new_datum' (line 534)
    [8208, 8256) 'base_datum' (line 534)
    [8288, 8336) 'new_datum' (line 534)
    [8368, 8416) 'base_datum' (line 534)
    [8448, 8496) 'new_datum' (line 534)
    [8528, 8576) 'base_datum' (line 534)
    [8608, 8656) 'new_datum' (line 534)
    [8688, 8736) 'base_datum' (line 534)
    [8768, 8816) 'new_datum' (line 534)
    [8848, 8896) 'base_datum' (line 534)
    [8928, 8976) 'new_datum' (line 534)
    [9008, 9056) 'base_datum' (line 534)
    [9088, 9136) 'new_datum' (line 534)
    [9168, 9216) 'base_datum' (line 536)
    [9248, 9296) 'new_datum' (line 536)
    [9328, 9376) 'base_datum' (line 536)
    [9408, 9456) 'new_datum' (line 536)
    [9488, 9536) 'base_datum' (line 536)
    [9568, 9616) 'new_datum' (line 536)
    [9648, 9696) 'base_datum' (line 536)
    [9728, 9776) 'new_datum' (line 536)
    [9808, 9856) 'base_datum' (line 536)
    [9888, 9936) 'new_datum' (line 536)
    [9968, 10016) 'base_datum' (line 536)
    [10048, 10096) 'new_datum' (line 536)
    [10128, 10176) 'base_datum' (line 536)
    [10208, 10256) 'new_datum' (line 536)
    [10288, 10336) 'base_datum' (line 536)
    [10368, 10416) 'new_datum' (line 536)
    [10448, 10496) 'base_datum' (line 536)
    [10528, 10576) 'new_datum' (line 536)
    [10608, 10656) 'base_datum' (line 536)
    [10688, 10736) 'new_datum' (line 536)
    [10768, 10816) 'base_datum' (line 538)
    [10848, 10896) 'new_datum' (line 538)
    [10928, 10976) 'base_datum' (line 538)
    [11008, 11056) 'new_datum' (line 538)
    [11088, 11136) 'base_datum' (line 538)
    [11168, 11216) 'new_datum' (line 538)
    [11248, 11296) 'base_datum' (line 538)
    [11328, 11376) 'new_datum' (line 538)
    [11408, 11456) 'base_datum' (line 538)
    [11488, 11536) 'new_datum' (line 538)
    [11568, 11616) 'base_datum' (line 538)
    [11648, 11696) 'new_datum' (line 538)
    [11728, 11776) 'base_datum' (line 538)
    [11808, 11856) 'new_datum' (line 538)
    [11888, 11936) 'base_datum' (line 538)
    [11968, 12016) 'new_datum' (line 538)
    [12048, 12096) 'base_datum' (line 538)
    [12128, 12176) 'new_datum' (line 538)
    [12208, 12256) 'base_datum' (line 538)
    [12288, 12336) 'new_datum' (line 538)
    [12368, 12416) 'base_datum' (line 540)
    [12448, 12496) 'new_datum' (line 540)
    [12528, 12576) 'base_datum' (line 540)
    [12608, 12656) 'new_datum' (line 540)
    [12688, 12736) 'base_datum' (line 540)
    [12768, 12816) 'new_datum' (line 540)
    [12848, 12896) 'base_datum' (line 540)
    [12928, 12976) 'new_datum' (line 540)
    [13008, 13056) 'base_datum' (line 540)
    [13088, 13136) 'new_datum' (line 540)
    [13168, 13216) 'base_datum' (line 540)
    [13248, 13296) 'new_datum' (line 540)
    [13328, 13376) 'base_datum' (line 540)
    [13408, 13456) 'new_datum' (line 540)
    [13488, 13536) 'base_datum' (line 540)
    [13568, 13616) 'new_datum' (line 540)
    [13648, 13696) 'base_datum' (line 540)
    [13728, 13776) 'new_datum' (line 540)
    [13808, 13856) 'base_datum' (line 540)
    [13888, 13936) 'new_datum' (line 540)
    [13968, 14016) 'dst_datum' (line 555)
    [14048, 14128) 'hll' (line 566) <== Memory access at offset 14048 is inside this variable
    [14160, 14240) 'bitmap' (line 571)
    [14272, 14392) 'percentile' (line 576)
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp and C++ exceptions *are* supported)
Thread T515 created by T0 here:
    #0 0x47aa972 in __interceptor_pthread_create ../../.././libsanitizer/asan/asan_interceptors.cpp:214
    #1 0xca19ed4 in __gthread_create /var/local/gcc/x86_64-pc-linux-gnu/libstdc++-v3/include/x86_64-pc-linux-gnu/bits/gthr-def
ault.h:663
    #2 0xca19ed4 in std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State>
 >, void (*)()) ../../../.././libstdc++-v3/src/c++11/[thread.cc:135](http://thread.cc:135/)
    #3 0x6cc2850 in starrocks::TaskWorkerPool::_spawn_callback_worker_thread(void* (*)(void*)) /root/starrocks/be/src/agent/ta
sk_worker_pool.cpp:279
    #4 0x6cc077f in starrocks::TaskWorkerPool::start() /root/starrocks/be/src/agent/task_worker_pool.cpp:165
    #5 0x6ca319c in starrocks::AgentServer::AgentServer(starrocks::ExecEnv*, starrocks::TMasterInfo const&) /root/starrocks/be
/src/agent/agent_server.cpp:83
    #6 0x61ce41b in starrocks::BackendService::BackendService(starrocks::ExecEnv*) /root/starrocks/be/src/service/backend_serv
ice.cpp:65
    #7 0x61ce571 in starrocks::BackendService::create_service(starrocks::ExecEnv*, int, starrocks::ThriftServer**) /root/starr
ocks/be/src/service/backend_service.cpp:68
    #8 0x4841613 in main /root/starrocks/be/src/service/starrocks_main.cpp:237
    #9 0x7f5bbc315554 in __libc_start_main (/lib64/libc.so.6+0x22554)

SUMMARY: AddressSanitizer: stack-use-after-scope /root/starrocks/be/src/storage/hll.h:93 in starrocks::HyperLogLog::HyperLogLo
g(starrocks::HyperLogLog const&)
Shadow bytes around the buggy address:
  0x0febd81a35c0: f2 f2 00 00 00 00 00 00 f2 f2 f2 f2 00 00 00 00
  0x0febd81a35d0: 00 00 f2 f2 f2 f2 00 00 00 00 00 00 f2 f2 f2 f2
  0x0febd81a35e0: 00 00 00 00 00 00 f2 f2 f2 f2 00 00 00 00 00 00
  0x0febd81a35f0: f2 f2 f2 f2 00 00 00 00 00 00 f2 f2 f2 f2 00 00
  0x0febd81a3600: 00 00 00 00 f2 f2 f2 f2 00 00 00 00 00 00 f2 f2
=>0x0febd81a3610: f2 f2[f8]f8 f8 f8 f8 f8 f8 f8 f8 f8 f2 f2 f2 f2
  0x0febd81a3620: 00 00 00 00 00 00 00 00 00 00 f2 f2 f2 f2 00 00
  0x0febd81a3630: 00 00 00 00 00 00 00 00 00 00 00 00 00 f3 f3 f3
  0x0febd81a3640: f3 f3 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0febd81a3650: 00 00 f1 f1 f1 f1 f1 f1 f8 f2 00 f2 f2 f2 00 f2
  0x0febd81a3660: f2 f2 00 f2 f2 f2 00 f2 f2 f2 00 f2 f2 f2 00 f2
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==3915==ABORTING
```
### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
 - MAIN-ASAN ba4e546

### PR DESCRIPTION
**Pull request by sevev**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #6123

## Problem Summary(Required) ：

When doing type convert of complex type, there is a dangling ptr which maybe cause corruption.

### CODE REVIEW

File: be/src/storage/schema_change.cpp [hunk 1]
```diff
@@ -642,6 +653,7 @@ bool ChunkSorter::sort(ChunkPtr& chunk, TabletSharedPtr new_tablet) {
     _swap_chunk = chunk->clone_empty_with_schema();
     _swap_chunk->append_selective(*chunk, selective.data(), 0, chunk->num_rows());
 
+    chunk->swap_chunk(*_swap_chunk);
```
Comment (sevev) [inline be/src/storage/schema_change.cpp:656]: This is another bug introduced by this pr https://github.com/StarRocks/starrocks/pull/4799/files

**Review by meegoo — DISMISSED** [review]

File: be/src/storage/schema_change.cpp (hunk 1 above)
Comment (trueeyu) [reply be/src/storage/schema_change.cpp:656]: Split PR

**Comment by trueeyu** [comment]
Please and test case

**Comment by sevev** [comment]
> Please and test case

Done,  add ut for it, PTAL again.

**Comment by sevev** [comment]
@trueeyu @chaoyli PTAL if you are available

**Review by chaoyli — APPROVED** [review]

**Review by trueeyu — APPROVED** [review]

**Pull request merged**